packages/cli/src/concurrency/concurrency-control.service.ts

```diff
@@ -0,0 +1,158 @@
+import { Logger } from '@/Logger';
+import config from '@/config';
+import { Service } from 'typedi';
+import { ConcurrencyQueue } from './concurrency-queue';
+import { UnknownExecutionModeError } from '@/errors/unknown-execution-mode.error';
+import { InvalidConcurrencyCapError } from '@/errors/invalid-concurrency-cap.error';
+import { ExecutionRepository } from '@/databases/repositories/execution.repository';
+import { License } from '@/License';
+import type { WorkflowExecuteMode as ExecutionMode } from 'n8n-workflow';
+
+@Service()
+export class ConcurrencyControlService {
+	private readonly isEnabled: boolean;
+
+	private readonly productionLimit: number;
+
+	private readonly productionQueue: ConcurrencyQueue;
+
+	constructor(
+		private readonly logger: Logger,
+		private readonly executionRepository: ExecutionRepository,
+		private readonly license: License,
+	) {
+		this.productionLimit = this.getProductionLimit();
+
+		if (this.productionLimit === 0) {
+			throw new InvalidConcurrencyCapError(this.productionLimit);
+		}
+
+		if (this.productionLimit < -1) {
+			this.productionLimit = -1;
+		}
+
+		if (this.productionLimit === -1 || config.getEnv('executions.mode') === 'queue') {
+			this.isEnabled = false;
+			this.log('Service disabled');
+			return;
+		}
+
+		this.productionQueue = new ConcurrencyQueue(this.productionLimit);
+
+		this.logInit();
+
+		this.isEnabled = true;
+
+		this.productionQueue.on('execution-throttled', async (executionId: string) => {
+			this.log('Execution throttled', { executionId });
+		});
+
+		this.productionQueue.on('execution-released', async (executionId: string) => {
+			this.log('Execution released', { executionId });
+			await this.executionRepository.resetStartedAt(executionId);
+		});
+	}
+
+	/**
+	 * Block or let through an execution based on concurrency capacity.
+	 */
+	async check({ mode, executionId }: { mode: ExecutionMode; executionId: string }) {
+		if (!this.isEnabled || this.isUncapped(mode)) return;
+
+		await this.productionQueue.enqueue(executionId);
+	}
+
+	/**
+	 * Release capacity back so the next execution in the production queue can proceed.
+	 */
+	release({ mode }: { mode: ExecutionMode }) {
+		if (!this.isEnabled || this.isUncapped(mode)) return;
+
+		this.productionQueue.dequeue();
+	}
+
+	/**
+	 * Remove an execution from the production queue, releasing capacity back.
+	 */
+	remove({ mode, executionId }: { mode: ExecutionMode; executionId: string }) {
+		if (!this.isEnabled || this.isUncapped(mode)) return;
+
+		this.productionQueue.remove(executionId);
+	}
+
+	/**
+	 * Remove many executions from the production queue, releasing capacity back.
+	 */
+	removeMany(executionIds: string[]) {
```

Comment:
I don't think this should ever happen.

One "fun fact" is that the UI won't let you delete executions that are running (the checkbox is not rendered for executions that are currently running, see below).

![Screenshot 2024-06-07 at 09 11 52](https://github.com/n8n-io/n8n/assets/219272/47c14d82-7fca-426a-a916-6d24ecd699eb)

This could only ever happen from the public API and it's also an error, we should not allow deletion of executions that are in progress.

The deletion should only be allowed for executions that are already settled, so this means that this method may be unnecessary unless we're changing this.

Reply:
[622eb47](https://github.com/n8n-io/n8n/pull/9453/commits/622eb47161944752b054644f1655e1d919e8932e)